src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+                assert isinstance(cursor, Block)
+                idx = fwd_sel(old_idx)
+
+            return evolve(old_path[:depth - 1] + [(attr, idx)] + old_path[depth:])
+
+        return forward
+
+
+@dataclass
+class Block(Cursor):
+    _path: list[tuple[str, Union[int, range]]]  # is range iff last entry
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (implementation)
+    # ------------------------------------------------------------------------ #
+
+    def parent(self) -> Node:
+        return Node(self._proc, self._path[:-1])
+
+    def before(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.start).before(dist)
+
+    def after(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.stop - 1).after(dist)
+
+    def prev(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes leading up to the start?
+        raise NotImplementedError('Block.prev')
+
+    def next(self, dist=1) -> Cursor:
+        # TODO: what should this mean?
+        #  1. The node after the block?
+        #  2. The block shifted over?
+        #  3. The block of nodes past the end?
+        raise NotImplementedError('Block.next')
+
+    # ------------------------------------------------------------------------ #
+    # Sequence interface implementation
+    # ------------------------------------------------------------------------ #
+
+    def __iter__(self):
+        attr, _range = self._path[-1]
+        block = self.parent()
+        for i in _range:
+            yield block.child(attr, i)
+
+    def __getitem__(self, i):
+        attr, r = self._path[-1]
+        r = r[i]
+        if isinstance(r, range):
+            if r.step != 1:
+                raise IndexError('block cursors must be contiguous')
+            return Block(self._proc, self._path[:-1] + [(attr, r)])
+        else:
+            return self.parent().child(attr, r)
+
+    def __len__(self):
+        _, _range = self._path[-1]
+        return len(_range)
+
+    # ------------------------------------------------------------------------ #
+    # AST mutation
+    # ------------------------------------------------------------------------ #
+
+    def _replace(self, nodes: list, *, empty_default=None):
+        """
+        This is an UNSAFE internal function for replacing a block in an AST with
+        a list of statements and providing a forwarding function as collateral.
+        It is meant to be package-private, not class-private, so it may be
+        called from other internal classes and modules, but not from end-user
+        code.
+        """
+        assert self._path
+        assert len(self) > 0
+
+        def update(node):
```

Comment:
maybe call this argument `parent` instead of `node` to make its role easier to guess?

Reply:
Might be good to similarly rename some other local variables named `node`.

Reply:
Yes, that's a good idea. I'll make that change before merging.